server/src/main/java/org/elasticsearch/common/document/DocumentField.java

```diff
@@ -113,10 +113,9 @@ public void writeTo(StreamOutput out) throws IOException {
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
         builder.startArray(name);
         for (Object value : values) {
-            // this call doesn't really need to support writing any kind of object.
-            // Stored fields values are converted using MappedFieldType#valueForDisplay.
-            // As a result they can either be Strings, Numbers, or Booleans, that's
-            // all.
+            // This call doesn't really need to support writing any kind of object, since the values
+            // here are always serializable to xContent. Each value could be a leaf types like a string,
+            // number, or boolean, a list of such values, or a map of such values with string keys.
```

Comment:
Ah! Except it won't be for stuff like `geo_points`, right?

Reply:
I do wonder if it'd be nicer to have the "type" of the field available so we don't need all the nasty reflection in `value`.

Reply:
But that is a problem for another time, I think.

Reply:
I agree, this is a really unfortunate part of the current PR. It's on the top of my mind in terms of what needs figuring out for a follow-up.